### PR TITLE
Add a more restrict CreateDownloadLink API

### PR DESCRIPTION
**Pull request by EngHabu**
Signed-off-by: Haytham Abuelfutuh <haytham@afutuh.com>

# TL;DR
Deprecate `CreateDownloadLocation` API to allow for a more controllable `CreateDownloadLink` API.
There are a few issues with the current API:
1. Admin is unable to link the request to an actual execution's output. It's, therefore, unable to enforce any authorization rules.
2. It exposes the entirety of the s3 bucket admin has access to, to be downloaded through the API with no meaningful Audit logs to record.
3. It assumes the deck/output file locations are known and not changeable which, I believe, isn't the right assumption. Ideally, even for outputs, users can just request the output location through the API. This makes the actual storage behind the output/decks/metadata completely portable.

## Type
 - [ ] Bug Fix
 - [X] Feature
 - [ ] Plugin

## Are all requirements met?

 - [X] Code completed
 - [X] Smoke tested
 - [X] Unit tests added
 - [ ] Code documentation added
 - [X] Any pending items have an associated Issue

## Tracking Issue
_Remove the '*fixes*' keyword if there will be multiple PRs to fix the linked issue_

fixes https://github.com/flyteorg/flyte/issues/3034

### CODE REVIEW

File: protos/flyteidl/service/dataproxy.proto [hunk 1]
```diff
@@ -63,6 +64,51 @@ message CreateDownloadLocationResponse {
...
+
+  // ExpiresIn defines a requested expiration duration for the generated url. The request will be rejected if this
+  // exceeds the platform allowed max.
+  // +optional. The default value comes from a global config.
+  google.protobuf.Duration expires_in = 5;
+}
+
+message CreateDownloadLinkResponse {
+  // SignedUrl specifies the url to use to download content from (e.g. https://my-bucket.s3.amazonaws.com/randomstring/suffix.tar?X-...)
+  string signed_url = 1;
```
Comment (kumare3) [inline protos/flyteidl/service/dataproxy.proto:106]: can we make this a list?

File: protos/flyteidl/service/dataproxy.proto [hunk 2]
```diff
@@ -63,6 +64,51 @@ message CreateDownloadLocationResponse {
   google.protobuf.Timestamp expires_at = 2;
 }
 
+enum ArtifactType {
+  ARTIFACT_TYPE_UNDEFINED = 0;
+  ARTIFACT_TYPE_DECK = 1;
+}
+
+message TaskExecutionId {
```
Comment (pingsutw) [inline protos/flyteidl/service/dataproxy.proto:72]: Should we use [TaskExecutionIdentifier](https://github.com/flyteorg/flyteidl/blob/70aea35519c8849c26932567bdfbaf8f8fc55055/protos/flyteidl/core/identifier.proto#L59-L65) in `flyteidl.core` instead?

Comment (EngHabu) [reply protos/flyteidl/service/dataproxy.proto:72]: oh yes! I didn't notice there is one. thank you!

File: protos/flyteidl/service/dataproxy.proto [hunk 3]
```diff
@@ -63,6 +64,51 @@ message CreateDownloadLocationResponse {
   google.protobuf.Timestamp expires_at = 2;
 }
 
+enum ArtifactType {
+  ARTIFACT_TYPE_UNDEFINED = 0;
+  ARTIFACT_TYPE_DECK = 1;
```
Comment (pmahindrakar-oss) [inline protos/flyteidl/service/dataproxy.proto:69]: Do we also need to  define ARTIFACT_TYPE_SOURCE which would be used for `pyflyte run --remote` and flytectl fast registration

File: protos/flyteidl/service/dataproxy.proto [hunk 4]
```diff
@@ -63,6 +64,45 @@ message CreateDownloadLocationResponse {
...
+
+  // ExpiresIn defines a requested expiration duration for the generated url. The request will be rejected if this
+  // exceeds the platform allowed max.
+  // +optional. The default value comes from a global config.
+  google.protobuf.Duration expires_in = 5;
+}
+
+message CreateDownloadLinkResponse {
+  // SignedUrl specifies the url to use to download content from (e.g. https://my-bucket.s3.amazonaws.com/randomstring/suffix.tar?X-...)
+  repeated string signed_url = 1;
```
Comment (katrogan) [inline protos/flyteidl/service/dataproxy.proto:100]: When does this return multiple values?

File: protos/flyteidl/service/dataproxy.proto [hunk 5]
```diff
@@ -75,11 +115,24 @@ service DataProxyService {
       description: "Creates a write-only http location that is accessible for tasks at runtime."
     };
   }
+
   // CreateDownloadLocation creates a signed url to download artifacts.
   rpc CreateDownloadLocation (CreateDownloadLocationRequest) returns (CreateDownloadLocationResponse) {
+    option deprecated = true;
```
Comment (katrogan) [inline protos/flyteidl/service/dataproxy.proto:121]: nit: should we also deprecate the CreateDownloadLocationRequest and CreateDownloadLocationResponse protos?

**Review by katrogan — DISMISSED** [review]

File: protos/flyteidl/service/dataproxy.proto [hunk 6]
```diff
@@ -63,6 +64,45 @@ message CreateDownloadLocationResponse {
...
+  oneof source {
+    // WorkflowId is the unique identifier for the workflow execution.
+    core.WorkflowExecutionIdentifier execution_id = 1;
+
+    // NodeId is the unique identifier for the node execution. For a task node, this will retrieve the output of the
+    // most recent attempt of the task.
+    core.NodeExecutionIdentifier node_id = 2;
+
+    // TaskId is the unique task execution identifier that points to a specific attempt of the task.
+    core.TaskExecutionIdentifier task_id = 3;
```
Comment (katrogan) [inline protos/flyteidl/service/dataproxy.proto:87]:
```suggestion
    core.TaskExecutionIdentifier task_execution_id = 3;
```

File: protos/flyteidl/service/dataproxy.proto [hunk 7]
```diff
@@ -63,6 +64,45 @@ message CreateDownloadLocationResponse {
...
+
+  // ARTIFACT_TYPE_DECK refers to the deck html file optionally generated after a task, a workflow or a launch plan
+  // finishes executing.
+  ARTIFACT_TYPE_DECK = 1;
+}
+
+message CreateDownloadLinkRequest {
+  oneof source {
+    // WorkflowId is the unique identifier for the workflow execution.
+    core.WorkflowExecutionIdentifier execution_id = 1;
```
Comment (katrogan) [inline protos/flyteidl/service/dataproxy.proto:80]: I see in the implementation we only support node execution identifiers - why include workflow and task execution identifiers in the proto definition at all?

**Review by wild-endeavor — COMMENTED** [review]
do we need retry number?

File: protos/flyteidl/service/dataproxy.proto (hunk 3 above)
Comment (pmahindrakar-oss) [reply protos/flyteidl/service/dataproxy.proto]: That would be used by CreateUploadRequest. Let me know what you think @EngHabu

**Review by pmahindrakar-oss — DISMISSED** [review]

File: protos/flyteidl/service/dataproxy.proto (hunk 4 above)
Comment (EngHabu) [reply protos/flyteidl/service/dataproxy.proto:100]: Ketan commented that... in case we start using this to download FlyteDirectory in the future,... etc.

**Comment by EngHabu** [comment]
> 

It's part of the taskexecution identifier

**Review by katrogan — APPROVED** [review]

**Review by wild-endeavor — APPROVED** [review]
we don't need inputs right?

**Comment by EngHabu** [comment]
right
